File: Computer_Sell_FrontEnd/src/app/services/stock.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CookieService } from 'ngx-cookie-service';
import { environment } from '../enviroment';

export type StockMovementType = 'IMPORT' | 'EXPORT' | 'RETURN' | 'ADJUSTMENT';

export interface StockHistoryResponse {
    id: string;
    productId: string;
    productName: string;
    movementType: StockMovementType;
    quantity: number;
    stockBefore: number;
    stockAfter: number;
    unitPrice: number;
    referenceId: string;
    referenceType: string;
    note: string;
    createdBy: string;
    movementDate: string;
    createdAt: string;
}

export interface StockAdjustmentRequest {
    productId: string;
    newStock: number;
    note?: string;
    createdBy: string;
}

export interface ResponseEnvelope<T = any> {
    code: number;
    message: string;
    data: T;
}

@Injectable({ providedIn: 'root' })
export class StockService {
    private readonly API = `${environment.apiUrl}/api/stock`;

    constructor(private http: HttpClient, private cookies: CookieService) { }

    // Lấy lịch sử kho theo sản phẩm
    getHistoryByProduct(productId: string): Observable<ResponseEnvelope<StockHistoryResponse[]>> {
        return this.http.get<ResponseEnvelope<StockHistoryResponse[]>>(
            `${this.API}/history/product/${productId}`,
            { headers: this.authHeaders() }
        );
    }

    // Lấy lịch sử kho theo loại biến động
    getHistoryByType(type: StockMovementType): Observable<ResponseEnvelope<StockHistoryResponse[]>> {
        return this.http.get<ResponseEnvelope<StockHistoryResponse[]>>(
            `${this.API}/history/type/${type}`,
            { headers: this.authHeaders() }
        );
    }

    // Lấy lịch sử kho theo khoảng thời gian
    getHistoryByDateRange(start: string, end: string): Observable<ResponseEnvelope<StockHistoryResponse[]>> {
        return this.http.get<ResponseEnvelope<StockHistoryResponse[]>>(
            `${this.API}/history/date-range`,
            {
                headers: this.authHeaders(),
                params: new HttpParams().set('start', start).set('end', end)
            }
        );
    }

    // Kiểm tra tồn kho hiện tại
    checkStock(productId: string): Observable<ResponseEnvelope<number>> {
        return this.http.get<ResponseEnvelope<number>>(
            `${this.API}/check/${productId}`,
            { headers: this.authHeaders() }
        );
    }

    // Kiểm tra tồn kho có đủ không
    checkStockAvailable(productId: string, quantity: number): Observable<ResponseEnvelope<boolean>> {
        return this.http.get<ResponseEnvelope<boolean>>(
            `${this.API}/check-available/${productId}`,
            {
                headers: this.authHeaders(),
                params: new HttpParams().set('quantity', quantity.toString())
            }
        );
    }

    // Điều chỉnh tồn kho (kiểm kê)
    adjustStock(request: StockAdjustmentRequest): Observable<ResponseEnvelope<null>> {
        return this.http.post<ResponseEnvelope<null>>(
            `${this.API}/adjust`,
            request,
            { headers: this.authHeaders() }
        );
    }

    private authHeaders(): HttpHeaders {
        const token = this.cookies.get('accessToken');
        return token ? new HttpHeaders({ Authorization: `Bearer ${token}` }) : new HttpHeaders();
    }
}

File: Computer-sell/src/main/java/com/trong/Computer_sell/service/StockService.java
package com.trong.Computer_sell.service;

import com.trong.Computer_sell.DTO.response.product.StockHistoryResponse;
import com.trong.Computer_sell.common.StockMovementType;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

/**
 * Service quản lý kho hàng
 */
public interface StockService {

    /**
     * Nhập kho - tăng tồn kho sản phẩm
     */
    void importStock(UUID productId, Integer quantity, BigDecimal unitPrice,
                     String referenceId, String referenceType, String note, String createdBy);

    /**
     * Xuất kho - giảm tồn kho sản phẩm (khi xác nhận đơn hàng)
     */
    void exportStock(UUID productId, Integer quantity, BigDecimal unitPrice,
                     String referenceId, String referenceType, String note, String createdBy);

    /**
     * Hoàn kho - tăng lại tồn kho (khi hủy đơn hàng)
     */
    void returnStock(UUID productId, Integer quantity, BigDecimal unitPrice,
                     String referenceId, String referenceType, String note, String createdBy);

    /**
     * Điều chỉnh kho - kiểm kê, sửa lỗi
     */
    void adjustStock(UUID productId, Integer newStock, String note, String createdBy);

    /**
     * Lấy lịch sử kho theo sản phẩm
     */
    List<StockHistoryResponse> getStockHistoryByProduct(UUID productId);

    /**
     * Lấy lịch sử kho theo loại biến động
     */
    List<StockHistoryResponse> getStockHistoryByType(StockMovementType type);

    /**
     * Lấy lịch sử kho theo khoảng thời gian
     */
    List<StockHistoryResponse> getStockHistoryByDateRange(LocalDateTime start, LocalDateTime end);

    /**
     * Kiểm tra tồn kho có đủ để xuất không
     */
    boolean checkStockAvailable(UUID productId, Integer quantity);

    /**
     * Lấy tồn kho hiện tại của sản phẩm
     */
    Integer getCurrentStock(UUID productId);
}
